### PR TITLE
#19 - Rich Text Renderer

### PR DESCRIPTION
**Pull request by jhancock532**
[Ticket on Codebase](https://projects.torchbox.com/projects/contentful-careers-page/tickets/19).

This adds a component that parses Contentful's rich text data into JSX elements.

### CODE REVIEW

File: components/RichText.tsx [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import { documentToReactComponents } from '@contentful/rich-text-react-renderer';
+import { BLOCKS, MARKS, INLINES } from '@contentful/rich-text-types';
+import React from 'react';
```
Comment (Morsey187) [inline components/RichText.tsx:3]: **Nickpick:** Keep react as the first import.
**Thought:** Depending on the tooling we should be okay to remove this import, see https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html and perhaps this relevant MR if any config changes are needed for linting: https://git.torchbox.com/sue-ryder/self-help/-/merge_requests/33/diffs#diff-content-dbc0c31823b8f2e4ed04a397722fed33a67f123f

File: components/RichText.tsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            return <p className={styles.paragraphLight}>{children}</p>;
+        case 'DARK':
+            return <p className={styles.paragraphDark}>{children}</p>;
+        case 'INDIGO':
+            return <p className={styles.paragraphIndigo}>{children}</p>;
+    }
+};
+
+const renderLink = (theme: Theme, node: any, children: React.ReactNode) => {
+    switch (theme) {
```
Comment (Morsey187) [inline components/RichText.tsx:25]: **Praise:** I really like your use of switch statements, you've actually influenced me to use them more.

File: components/RichText.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                <span className={styles.bold}>{text}</span>
+            ),
+            [MARKS.ITALIC]: (text: React.ReactNode) => (
+                <span className={styles.italic}>{text}</span>
+            ),
+            [MARKS.UNDERLINE]: (text: React.ReactNode) => (
+                <span className={styles.underline}>{text}</span>
+            ),
+        },
+        renderNode: {
```
Comment (Morsey187) [inline components/RichText.tsx:60]: Sorry James, I'll have to review this another time to make sure I understand it properly before giving feedback, will look into this again tomorrow when i have more time.

File: components/RichText.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
+import { documentToReactComponents } from '@contentful/rich-text-react-renderer';
+import { BLOCKS, MARKS, INLINES } from '@contentful/rich-text-types';
+import React from 'react';
+import styles from 'styles/RichText.module.scss';
```
[the remark this answers is not in the document]
Comment (jhancock532) [reply components/RichText.tsx:4]: This makes a lot more sense. I didn't do this early on as it gave a compiling error, but now I've updated the webpack config it works as expected

File: components/RichText.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                </a>
+            );
+    }
+};
+
+const getRenderOptions = (theme: Theme) => {
+    return {
+        renderMark: {
+            [MARKS.BOLD]: (text: React.ReactNode) => (
+                <span className={styles.bold}>{text}</span>
```
[the remark this answers is not in the document]
Comment (jhancock532) [reply components/RichText.tsx:51]: I decided against these tags originally as I favoured only using CSS for styling, but semantic HTML makes more sense accessibility wise

File: components/RichText/RichText.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                return <h3 className={styles.h3}>{children}</h3>;
+            },
+            [BLOCKS.PARAGRAPH]: (
+                node: React.ReactNode,
+                children: React.ReactNode,
+            ) => renderParagraph(theme, children),
+            // Note that TypeScript expects this function to return an object with functions that take in `React.ReactNode`s
+            // In reality, JSON data is being passed into this function, hence the `:any` type here
+            [INLINES.HYPERLINK]: (node: any, children: React.ReactNode) =>
+                renderLink(theme, node, children),
```
Comment (thibaudcolas) [inline components/RichText/RichText.tsx:82]:
```suggestion
            [INLINES.HYPERLINK]: (node: Block | Inline, children: React.ReactNode) =>
                renderLink(theme, (node as Hyperlink), children),
```

**Review by thibaudcolas — APPROVED** [review]
For consideration – doesn’t necessarily need changing.

**Pull request merged**